### PR TITLE
HSEARCH-3615 Document the query DSL in details

### PR DESCRIPTION
**Pull request by yrodiere**
[HSEARCH-3615](https://hibernate.atlassian.net/browse/HSEARCH-3615): Document the query DSL in details

~Based on #2069 , which should be merged first.~ => Done

### CODE REVIEW

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 1]
```diff
@@ -230,24 +360,168 @@ include::todo-placeholder.asciidoc[]
...
+you will need information about the exact query being executed,
+and about the index content.
+
+To gain insight about what ends up being executed exactly,
+one option is to <<search-dsl-query-object,create a `SearchQuery` object>>
+using `toQuery()` at the end of the query definition,
+then call `toString()` to get a String representation of that query.
+
+Another option is to take advantage of logs:
+all executed search queries are logged to the log category `org.hibernate.search.fulltext_query`
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: I have a doubt here: is the dash after `fulltext` correct?

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 2]
```diff
@@ -230,24 +360,168 @@ include::todo-placeholder.asciidoc[]
...
+and then use the backend-specific `explain` methods;
+the result of these methods will explain how the score of a specific document was computed.
+See below for examples.
 
-// TODO https://docs.jboss.org/hibernate/search/5.11/reference/en-US/html_single/#_understanding_results
-// TODO toString(), getQueryString(), and logging, too?
+To retrieve an explanation for all matches in one call,
+`explanation` projections are available:
+see <<search-dsl-projection-extensions-lucene-explanation,here for Lucene>>
+and <<search-dsl-projection-extensions-elasticsearch-explanation,here for Elasticsearch>>
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: I think we need a dot at the end of the line.

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 3]
```diff
@@ -67,14 +67,154 @@ to group hits and compute aggregated metrics for each group -- hit count by cate
...
+[[search-dsl-query-fetching-results-pagination]]
+==== Pagination
+// Search 5 anchors backward compatibility
+[[_pagination]]
+
+Pagination is the concept of splitting hits in successive "pages",
+all pages containing a fixed number of elements (except potentially the last one).
+When displaying results on a web page,
+the user will be able to go to an arbitrary page and
+
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/search-dsl.asciidoc]: An oversight. Thanks for the heads-up.

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 4]
```diff
@@ -230,24 +360,168 @@ include::todo-placeholder.asciidoc[]
...
+without bringing anything worthwhile.
+
+However, in some cases a query object can be useful.
+To get a query object, just call `toQuery()` at the end of the query definition:
+
+.Getting a `SearchQuery` object
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/query/QueryDslIT.java[tags=searchQuery]
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/search-dsl.asciidoc]: Right, done.

File: documentation/src/main/asciidoc/search-dsl.asciidoc (hunk 1 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/search-dsl.asciidoc]: You're right, I was using the log category from Search 5. In Search 6, it's simply `org.hibernate.search.query`.

**Comment by yrodiere** [comment]
Thanks! I addressed all of your comments and force-pushed. Waiting for CI, then I'll merge.

**Comment by yrodiere** [comment]
Merged, thanks!

**Pull request merged**